source/common/formatter/substitution_formatter.cc

```diff
@@ -685,431 +685,430 @@ class StreamInfoSslConnectionInfoFieldExtractor : public StreamInfoFormatter::Fi
+                   return connection_info.issuerPeerCertificate();
+                 });
+           }},
+          {"DOWNSTREAM_PEER_CERT",
+           []() {
+             return std::make_unique<StreamInfoSslConnectionInfoFieldExtractor>(
+                 [](const Ssl::ConnectionInfo& connection_info) {
+                   return connection_info.urlEncodedPemEncodedPeerCertificate();
+                 });
+           }},
+          {"UPSTREAM_TRANSPORT_FAILURE_REASON",
+           []() {
+             return std::make_unique<StreamInfoStringFieldExtractor>(
+                 [](const StreamInfo::StreamInfo& stream_info) {
+                   absl::optional<std::string> result;
+                   if (stream_info.upstreamInfo().has_value() &&
+                       !stream_info.upstreamInfo()
+                            .value()
+                            .get()
+                            .upstreamTransportFailureReason()
+                            .empty()) {
+                     result =
+                         stream_info.upstreamInfo().value().get().upstreamTransportFailureReason();
+                   }
+                   return result;
+                 });
+           }},
+          {"HOSTNAME",
+           []() {
+             absl::optional<std::string> hostname = SubstitutionFormatUtils::getHostname();
+             return std::make_unique<StreamInfoStringFieldExtractor>(
+                 [hostname](const StreamInfo::StreamInfo&) { return hostname; });
+           }},
+          {"FILTER_CHAIN_NAME",
+           []() {
+             return std::make_unique<StreamInfoStringFieldExtractor>(
+                 [](const StreamInfo::StreamInfo& stream_info) -> absl::optional<std::string> {
+                   if (!stream_info.filterChainName().empty()) {
+                     return stream_info.filterChainName();
+                   }
+                   return absl::nullopt;
+                 });
+           }},
+          {"VIRTUAL_CLUSTER_NAME",
+           []() {
+             return std::make_unique<StreamInfoStringFieldExtractor>(
+                 [](const StreamInfo::StreamInfo& stream_info) -> absl::optional<std::string> {
+                   return stream_info.virtualClusterName();
+                 });
+           }},
+          {"TLS_JA3_FINGERPRINT", []() {
+             return std::make_unique<StreamInfoStringFieldExtractor>(
+                 [](const StreamInfo::StreamInfo& stream_info) {
+                   absl::optional<std::string> result;
+                   if (!stream_info.downstreamAddressProvider().ja3Hash().empty()) {
+                     result = std::string(stream_info.downstreamAddressProvider().ja3Hash());
+                   }
+                   return result;
+                 });
+           }}});
+}
 
 StreamInfoFormatter::StreamInfoFormatter(const std::string& field_name) {
-  auto it = known_field_extractors_.find(field_name);
+  auto it = getKnownFieldExtractors().find(field_name);
```

Comment:
make a temp for the table.
```
  const FieldExtractorLookupTbl& extractors = getKnownFieldExtractors();
  auto it = extractors.find(...);
   ...
``

Reply:
to give color to this issue: it's (IMO) a bad pattern to call a function to return a collection to call find, and then call the function *again* to get the collection to get the end() iterator.

Looking at that code makes me worry whether the function will return the same collection each time. I know it won't in this case, but it's better not to have that pattern. 

It's also slightly faster to take the temp because we do the compiler-generated thread-safe lazy-init only once.

Reply:
Good point. Changed it as you suggested. Thanks! I always missed something like map<>::npos, which would allow for checking if iterator is valid without referencing the container's pointer ot reference.